02-peer-protocol.md

```diff
@@ -560,6 +1047,309 @@ ones it has sent so it can use them should they be requested by
 incoming HTLCs.  The recipient only need remember one, for use in
 `r` route hints in BOLT 11 invoices.
 
+If an RBF negotiation is in progress when a `channel_ready` message is
+exchanged, the negotiation must be abandoned.
+
+## Channel Establishment v2
+
+This is a revision of the channel establishment protocol.
+It changes the previous protocol to allow the `accept_channel2` peer
+(the *accepter*/*non-initiator*) to contribute inputs to the funding
+transaction, via the interactive transaction construction protocol.
+
+        +-------+                              +-------+
+        |       |--(1)--- open_channel2  ----->|       |
+        |       |<-(2)--- accept_channel2 -----|       |
+        |       |                              |       |
+    --->|       |      <tx collaboration>      |       |
+    |   |       |                              |       |
+    |   |       |--(3)--  commitment_signed -->|       |
+    |   |       |<-(4)--  commitment_signed ---|       |
+    |   |   A   |                              |   B   |
+    |   |       |<-(5)--  tx_signatures -------|       |
+    |   |       |--(6)--  tx_signatures ------>|       |
+    |   |       |                              |       |
+    |   |       |--(a)--- tx_init_rbf -------->|       |
+    ----|       |<-(b)--- tx_ack_rbf ----------|       |
+        |       |                              |       |
+        |       |    <tx rbf collaboration>    |       |
+        |       |                              |       |
+        |       |--(c)--  commitment_signed -->|       |
+        |       |<-(d)--  commitment_signed ---|       |
+        |       |                              |       |
+        |       |<-(e)--  tx_signatures -------|       |
+        |       |--(f)--  tx_signatures ------>|       |
+        |       |                              |       |
+        |       |--(7)--- channel_ready  ----->|       |
+        |       |<-(8)--- channel_ready  ------|       |
+        +-------+                              +-------+
+
+        - where node A is *opener*/*initiator* and node B is
+          *accepter*/*non-initiator*
+
+### The `open_channel2` Message
+
+This message initiates the v2 channel establishment workflow.
+
+1. type: 64 (`open_channel2`)
+2. data:
+   * [`chain_hash`:`chain_hash`]
+   * [`channel_id`:`temporary_channel_id`]
+   * [`u32`:`funding_feerate_perkw`]
+   * [`u32`:`commitment_feerate_perkw`]
+   * [`u64`:`funding_satoshis`]
+   * [`u64`:`dust_limit_satoshis`]
+   * [`u64`:`max_htlc_value_in_flight_msat`]
+   * [`u64`:`htlc_minimum_msat`]
+   * [`u16`:`to_self_delay`]
+   * [`u16`:`max_accepted_htlcs`]
+   * [`u32`:`locktime`]
+   * [`point`:`funding_pubkey`]
+   * [`point`:`revocation_basepoint`]
+   * [`point`:`payment_basepoint`]
+   * [`point`:`delayed_payment_basepoint`]
+   * [`point`:`htlc_basepoint`]
+   * [`point`:`first_per_commitment_point`]
```

Comment:
I think I made a mistake in #25 (84 years ago), I should have also included the `second_per_commitment_point` in `open_channel` (we currently send later, in the `channel_ready`).

It doesn't change anything from a protocol point of view, but it would make the implementation easier, at least in eclair. We currently have to make a special case for the first commitments and use a placeholder value, because during normal channel operations we always have the next `commitment_point` available when we don't have a sig in-flight. Curious if other implementers ran into this?

Maybe we could take the opportunity to add it here.

Reply:
to check my understanding: this is a proposal to send both the `first` and `second` commitment point in `open_channel2`. The reason for this is so that the in-program data structures are correctly set with both when instantiated (with both current + next points).

Currently we send this second point in `channel_ready`, which happens at some (indeterminate) time after the channel has been committed to chain. 

Seems reasonable to me.

Might impact the `splicing` proposal, cc @ddustin ?

Reply:
Given that the `channel_ready` message is re-used between the v1 and v2 protocols though, I don't think we can remove it from the message. So sending it in `open_channel2` would be redundant.

Reply:
> to check my understanding [...]

Your understanding is correct

I actually came to this while working on a `splicing` prototype. The natural thing to do to confirm a splice is to send a `channel_ready`, but then `channel_ready.second_per_commitment_point` is odd. This reminded me that this field shouldn't have been there in the first place.

I suppose @ddustin has run into this too? Maybe by reusing `channel_ready` isn't such a good idea if we need to introduce a new equivalent message for `splicing`.

Reply:
@niftynei this should also be added to `accept_channel2`. I agree that we cannot change `channel_ready` at this point, but it's ok, we can simply ignore its `next_per_commitment_point` when using dual funding. I don't think it's worth introducing a new message just for this. Maybe once we deprecate v1 open we'll have ideas to just re-purpose that field?

Reply:
> I actually came to this while working on a `splicing` prototype. The natural thing to do to confirm a splice is to send a `channel_ready`, but then `channel_ready.second_per_commitment_point` is odd. This reminded me that this field shouldn't have been there in the first place.
> 
> I suppose @ddustin has run into this too? Maybe reusing `channel_ready` isn't such a good idea if we need to introduce a new equivalent message for `splicing`.

Yup the splicing spec calls for a `splice_locked` message but doesn't yet define it. For development I've been using this message spec.

```
splice_locked(78)
+ [channel_id]channel_id
```

Reminds me we should add it explicitly into the spec. I think `channel_id` is all it needs although it may be worth thinking about the situation where we have multiple RBF competing splices and saying which one confirmed. Both nodes should really be monitoring for all of them already though 🤷